src/Ansatz/Chain.jl

```diff
@@ -144,6 +146,118 @@ function rightindex(::Union{Open,Periodic}, tn::Chain, site::Site)
     end
 end
 
+"""
+    adjoint(tn::TensorNetwork{<:Quantum})
+
+Return the adjoint [`TensorNetwork`](@ref).
+
+# Implementation details
+
+The tensors are not transposed, just `conj!` is applied to them.
+"""
+function Base.adjoint(qtn::Chain)
+    qtn = deepcopy(qtn)
+
+    foreach(conj!, tensors(qtn))
+
+    return qtn
+end
+
+function LinearAlgebra.norm(ψ::Chain, p::Real = 2; kwargs...)
+    p != 2 && throw(ArgumentError("p=$p is not implemented yet"))
+
+    # TODO: Replace with hcat when implemented
+    return contract(tensors(TensorNetwork(ψ))..., tensors(TensorNetwork(ψ'))...; kwargs...) |> only |> sqrt |> abs
```

Comment:
This is gonna fail because inner indices are unchanged so you will form hyperindices. Currently, it's better if you canonize to the first site and just contract that tensor with itself conjugated.

Add a `TODO` message to say that we need to support just direct contraction in the future.

Reply:
Okay, but I just want to point out that is currently not failing.

Reply:
It's not failing because I guess the tensors are left-canonical and thus, the give the identity even when contracted with virtual hyperindices.

The function is not going to crash, but the output will be wrong.

Reply:
I don't like the idea of canonizing since we have not yet implemented it for `MPO`, can't we just make that we first contract each `psi` and `adjoint(psi)` by separate and then contract them together? I know this is not optimal in the long run but it will work for all `Chain` now.

Reply:
That won't work because the virtual indices will overlap. The easiest thing to do is to reindex the virtual indices of the adjoint TN.

Reply:
No, because if you contract all from `psi` you get only the physical indices, and the same happens with `adjoint(psi)`, so then you are left with a contraction of two tensors of size `n` (where `n` is the number of tensors).

Reply:
But this increases memory exponentially! I would go for reindexing the virtual indices, then calling `merge` on the two TNs and finally, `contract`.

Reply:
Okay, then should this reindexing go directly into `adjoint`?

Reply:
mmm okay. Maybe I will move the reindexing to another part but for now is ok. Remember to reindex all indices that are not the physical indices. Do not use `inds(tn, :inner)`. The reason is that there might be hyperindices after some canonization and `inds(tn, :inner)` only selects normal edges (not hyperindices).

Reply:
Okay, done!